### ISSUE TITLE
False positive unbalanced-tuple-unpacking for scipy.optimize.curve_fit with pylint 2.5

### ISSUE THREAD

**Issue by willsALMANJ** [PyCQA/pylint#3601]
### Steps to reproduce
Run pylint 2.5 on the following file:

```
"mwe curve_fit unbalanced-tuple-unpacking"
import numpy as np
from scipy.optimize import curve_fit


def main():
    "main"
    xvec = np.linspace(0, 1)
    yvec = 3 * xvec
    popt, pcov = curve_fit(lambda x, a: a * x, xvec, yvec)
    print("popt")
    print(popt)
    print("pcov")
    print(pcov)


if __name__ == "__main__":
    main()
```

### Current behavior
pylint 2.5 gives the warning:

```
tmp.py:10:4: W0632: Possible unbalanced tuple unpacking with sequence defined at line 811 of scipy.optimize.minpack: left side has 2 label(s), right side has 5 value(s) (unbalanced-tuple-unpacking
```

for this file.

### Expected behavior
pylint 2.4.0 gives no warnings for the example file. I think this is the expected behavior.

Looking at the source code in scipy ([here](https://github.com/scipy/scipy/blob/v1.4.1/scipy/optimize/minpack.py#L810-L813)), there are two return statements with different numbers of parameters depending on the input parameters. I think pylint should be able to figure out that for the example above the two parameter return statement will be executed so there is no unbalanced tuple unpacking issue.

### pylint --version output
```
pylint 2.5.2
astroid 2.4.1
Python 3.8.2 | packaged by conda-forge | (default, Apr 24 2020, 08:20:52) 
[GCC 7.3.0]
```

**Comment by AWhetter** [comment]
Given the complexity of the method it's incredibly difficult (maybe even impossible) to determine what will be returned without actually running the code. The warning is correct in that it's "possible" that the tuple unpacking will fail depending on what happens at run time. To me this feels more like poor API design on scipy's part, but there's nothing that we can do to fix this problem.

**Comment by willsALMANJ** [comment]
Fair enough. Certainly the API is poor. I see your point that any unguarded unpacking of the return value should be treated as invalid since the function can return more than one length of tuple and determining the length with static analysis is too complex. Though I also think it wouldn't be bad if pylint gave no warning when unpacking to either two or five elements. Somehow pylint determined that the function returned a five member tuple and that same analysis should be able to see the two member tuple return statement as well.

I am not that familiar with how pylint works, but I get no error for this code, so it seems like pylint does not always warn about unbalanced unpacking for variable length return values:

```python
"mwe"
import random


def args():
    "args"
    return tuple([0] * random.randrange(3))


def main():
    "main"
    arg1, arg2 = args()
    print(arg1, arg2)


if __name__ == "__main__":
    main()
```

For my own purposes, I will use this form for the original example to avoid adding a pylint exception:

```python
"mwe curve_fit unbalanced-tuple-unpacking"
import numpy as np
from scipy.optimize import curve_fit


def main():
    "main"
    xvec = np.linspace(0, 1)
    yvec = 3 * xvec
    fit_result = curve_fit(lambda x, a: a * x, xvec, yvec)
    popt, pcov = fit_result[:2]
    print("popt")
    print(popt)
    print("pcov")
    print(pcov)


if __name__ == "__main__":
    main()
```

Thinking about your point though, I think there is room for improving pylint's analysis even if the preference is that pylint always warn about unbalanced unpacking when a function can return more than one length tuple. pylint treats the curve_fit as though it always returns a five member tuple. There is no warning when the code unpacks to five elements even if the function is called in a way that only returns two. So a more robust behavior would be for pylint to notice that there are multiple return statements and to infer that the function returns a variable length tuple and warn for any unguarded unpacking.
